### ISSUE TITLE
Promise-type checking gives false errors

### ISSUE THREAD

**Issue by iislucas** [Microsoft/TypeScript#2010]
The following code used to type-check fine (Using the types for Promises prior to: https://github.com/borisyankov/DefinitelyTyped/commit/c95561be3526f90f0e717734a659d1c01d0492d4 ), but since they now use types based on:  https://github.com/Microsoft/TypeScript/blob/master/src/lib/es6.d.ts#L3542 
the code produces false type errors. 
```ts
/// <reference path='es6-promise.d.ts' />

interface T1 {
  __t1: string;
}

interface T2 {
  __t2: string;
}

interface T3 {
  __t3: string;
}

function f1() : Promise<T1> {
  return Promise.resolve({ __t1: "foo_t1" });
}

function f2(x:T1) : T2 {
  return { __t2: x.__t1 + ":foo_21" };
}

var x3 = f1()
  .then(f2, (e:Error) => {
    console.log("error 1");
    throw e;
  })
  .then((x:T2) => {
    return { __t3: x.__t2 + "bar" };
  });
```
It gives these errors (in TypeScript 1.4.1): 
```
promise-oddity.ts(23,10):  error TS2453: 
The type argument for type parameter 'U' cannot be inferred from the usage. 
Consider specifying the type arguments explicitly.
  Type argument candidate 'T2' is not a valid type argument 
because it is not a supertype of candidate 'void'.
```

**Comment by RyanCavanaugh** [comment]
Is this still reproing with the latest from master? I think we made some fixes in this area

**Comment by iislucas** [comment]
Just retested with typescript 1.5-beta and it is still broken in that version.

**Comment by iislucas** [comment]
Just retested with latest from DefinitelyTyped (https://github.com/borisyankov/DefinitelyTyped/commit/7f9ae75f8c8b2d0ae92103c36ef835c488271ac8#diff-f53514e542177d42bff33f6076b1fea9) and latest typescript (1.5-beta), and there is still the same problem.

**Comment by mhegazy** [comment]
@JsonFreeman can you take a look.

**Comment by JsonFreeman** [comment]
The problem is that the error callback is void, and we try to make an inference from that void return type. You can tell because it works if you return something from your callback, like changing it to:
```ts
(e:Error) => {
    console.log("error 1");
    throw e;
    var t2: T2;
    return t2;
  }
```

I believe the correct fix is to add another overload to then (after the existing overload), that looks like this:
```ts
then<TResult>(onfulfilled?: (value: T) => TResult | PromiseLike<TResult>, onrejected?: (reason: any) => void): Promise<TResult>;
```
Note that this accounts for a void error function, but does not try to infer from it.

**Comment by mhegazy** [comment]
@JsonFreeman do you see the same issue with the promise typings in lib.es6.d.ts?

**Comment by JsonFreeman** [comment]
I thought that was what we were talking about. That is exactly where I found the issue.

**Comment by mhegazy** [comment]
The issue was reported with borisyankov/DefinitelyTyped@7f9ae75#diff-f53514e542177d42bff33f6076b1fea9

Which is slightly different from the one in lib.es6.d.ts

if the issue exists in both we need to fix them.

**Comment by JsonFreeman** [comment]
Yes, I see it in both. Will fix.

**Comment by iislucas** [comment]
Still waiting on DefinitelyTyped fix too, right?

**Comment by JsonFreeman** [comment]
Yes, https://github.com/borisyankov/DefinitelyTyped/pull/4280. Not sure how long it normally takes though.

**Comment by basarat** [comment]
> Still waiting on DefinitelyTyped fix too, right?

merged :rose:
